onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -91,9 +94,70 @@ Status AttentionBase::CheckInputs(const Tensor* input,
     }
   }
 
+  if (past != nullptr) {  // past is optional
+    if (!is_unidirectional_) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 4 (past) is only allowed for unidirectional");
+    }
+
+    const auto past_dims = past->Shape().GetDims();
+    if (past_dims.size() != 5) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 4 is expected to have 5 dimension, got ",
+                             past_dims.size());
+    }
+    if (static_cast<int>(past_dims[0]) != 2) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Inputs 4 dimension 0 shall have length of 2");
+    }
+    if (static_cast<int>(past_dims[1]) != batch_size) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Inputs 4 dimension 1 shall have same length as dimension 0 of input 0");
+    }
+    if (static_cast<int>(past_dims[2]) != num_heads_) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Inputs 4 dimension 2 shall have length of num_heads", num_heads_);
+    }
+    if (static_cast<int>(past_dims[4]) != hidden_size / num_heads_) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Inputs 4 dimension 2 shall have length of ", hidden_size / num_heads_);
+    }
+  }
+
   return Status::OK();
 }
 
+Tensor* AttentionBase::GetPresent(OpKernelContext* context,
+                                  const Tensor* past,
+                                  int batch_size,
+                                  int head_size,
+                                  int sequence_length,
+                                  int& past_sequence_length) const {
+  // Input and output shapes:
+  //   past        : (2, batch_size, num_heads, past_sequence_length, head_size)
+  //   present     : (2, batch_size, num_heads, past_sequence_length + sequence_length, head_size)
+
+  std::vector<int64_t> present_dims;
+  if (nullptr != past) {
+    const auto past_dims = past->Shape().GetDims();
+    past_sequence_length = static_cast<int>(past_dims[3]);
+
+    present_dims.push_back(past_dims[0]);
+    present_dims.push_back(past_dims[1]);
+    present_dims.push_back(past_dims[2]);
+    present_dims.push_back(static_cast<int64_t>(past_sequence_length + sequence_length));
+    present_dims.push_back(past_dims[4]);
+  } else {
+    present_dims.push_back(static_cast<int64_t>(2));
+    present_dims.push_back(static_cast<int64_t>(batch_size));
+    present_dims.push_back(static_cast<int64_t>(num_heads_));
+    present_dims.push_back(static_cast<int64_t>(sequence_length));
+    present_dims.push_back(static_cast<int64_t>(head_size));
+  }
+
```

Comment:
looks like we can make the code more concise with initializer_list:  
std::vector<int64_t> present_dims{2, batch_size, num_heads_, sequence_length, head_size}; 
  if (nullptr != past) {
    const auto past_dims = past->Shape().GetDims();
    past_sequence_length = static_cast<int>(past_dims[3]);
    present_dims[3] += past_sequence_length; 
}